File: tests/src/org.joda/joda-convert/2.2.4/src/test/java/org_joda/joda_convert/TypesTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_joda.joda_convert;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertSame;

import java.lang.reflect.Method;
import java.lang.reflect.ParameterizedType;

import org.joda.convert.StringConvert;
import org.junit.jupiter.api.Test;

public class TypesTest {
    @Test
    public void createsArrayClassForComponentType() throws Exception {
        Method getArrayClass = Class.forName("org.joda.convert.Types")
                .getDeclaredMethod("getArrayClass", Class.class);
        getArrayClass.setAccessible(true);

        assertSame(String[].class, getArrayClass.invoke(null, String.class));
    }

    @Test
    public void formatsParsedParameterizedTypes() {
        StringConvert convert = StringConvert.create();
        ParameterizedType type = convert.convertFromString(
                ParameterizedType.class,
                "java.util.Map<java.lang.String, java.util.List<java.lang.Integer>>");

        assertEquals(
                "java.util.Map<java.lang.String, java.util.List<java.lang.Integer>>",
                convert.convertToString(ParameterizedType.class, type));
    }
}

File: tests/src/org.playframework.anorm/anorm_3/2.8.1/src/test/scala/org_playframework_anorm/anorm_3/TypesTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_playframework_anorm.anorm_3

import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method
import java.util.concurrent.Callable

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertSame
import org.junit.jupiter.api.Test

class TypesTest {
  @Test
  def createsArrayClassesForComponentClasses(): Unit = {
    val getArrayClass: Method = typesClass.getDeclaredMethod("getArrayClass", classOf[Class[?]])
    getArrayClass.setAccessible(true)

    val arrayClass: AnyRef = getArrayClass.invoke(null, classOf[String])

    assertSame(classOf[Array[String]], arrayClass)
  }

  @Test
  def createsInterfaceProxiesWithSuppliedInvocationHandlers(): Unit = {
    val newProxy: Method = typesClass.getDeclaredMethod(
      "newProxy",
      classOf[Class[?]],
      classOf[InvocationHandler]
    )
    newProxy.setAccessible(true)
    val handler: InvocationHandler = (proxy: Object, method: Method, args: Array[Object]) => {
      method.getName match {
        case "call" => "from-proxy"
        case "toString" => "TypesTest proxy"
        case "hashCode" => Integer.valueOf(System.identityHashCode(proxy))
        case "equals" => Boolean.box(proxy eq args.apply(0))
        case other => throw new UnsupportedOperationException(other)
      }
    }

    val proxy: Callable[String] = newProxy
      .invoke(null, classOf[Callable[?]], handler)
      .asInstanceOf[Callable[String]]

    assertEquals("from-proxy", proxy.call())
  }

  private def typesClass: Class[?] = Class.forName("org.joda.convert.Types")
}
